Scripts/Tests/xGamePythonReview.py

```diff
@@ -0,0 +1,190 @@
+import inspect
+import io
+import os
+from pathlib import Path
+import sys
+import traceback
+from typing import *
+
+# Environment variables are being used so that you can copy this outside of the workflow
+# and trivually run the tests yourself.
+repo_path = Path(os.environ['PLASMA_PATH'])
+input_path = repo_path.joinpath("Scripts", "Python")
+
+# Setup PYTHONPATH like Plasma does in internal builds.
+sys.path.append(str(input_path))
+sys.path.append(str(input_path.joinpath("plasma")))
+
+# Results
+found_ids: Dict[int, Set[Path]] = defaultdict(set)
+exitcode = 0
+
+# Baseline format string for a parsable thingy.
+result_str = "---{status} {file}:{line} {message}---\n"
+
+def dump_exc(what: str, output: io.IOBase = sys.stderr) -> str:
+    global exitcode
+    exitcode = 1
+
+    # Because format_exc() and print_exc() don't give the same result.
+    buf = io.StringIO()
+    traceback.print_exc(file=buf)
+    result = buf.getvalue().rstrip()
+    print(f"... Blast! Error during {what}:", result, sep="\n")
+
+    # Parsable result
+    exc_type, exc_value, exc_tb = sys.exc_info()
+
+    # For some reason, Python 3.8 uses the name of THIS file first, so let's just walk the trace
+    # until we get something inside the scripts path.
+    for frame, line in traceback.walk_tb(exc_tb):
+        naughty_path = Path(frame.f_code.co_filename)
+        if not naughty_path.is_absolute():
+            naughty_path = repo_path.joinpath(naughty_path)
+
+        # Python 3.8 lacks is_relative_to(). dang, so fake it with this goofy construct.
+        try:
+            naughty_path.relative_to(input_path)
+        except ValueError:
+            continue
+        else:
+            if naughty_path.exists():
+                break
+
+    try:
+        error_path = naughty_path.relative_to(repo_path)
+    except ValueError:
+        output.write("Crap, I couldn't figure this one out!\n")
+        output.write(result)
+        output.write("\n")
+        sys.exit(1)
+
+    output.write(
+        result_str.format(
+            status="ERROR",
+            file=error_path,
+            line=line,
+            message=f"{exc_type.__name__}: {exc_value}"
+        )
+    )
+
+    return result
+
+def dump_missing_id(filename: Path, output: io.IOBase = sys.stderr):
+    global exitcode
+    exitcode = 1
+
+    output.write(
+        result_str.format(
+            status="WARNING",
+            file=filename.relative_to(repo_path),
+            line="",
+            message="PythonFileMod(TM) missing an `id` attribute"
+        )
+    )
+
+def dump_dupe_id(output: io.IOBase = sys.stderr):
+    global exitcode
+
+    for id, dupe_files in get_duplicated_ids().items():
+        exitcode = 1
+        for py_file in dupe_files:
+            other_dupes = dupe_files - set((py_file,))
+            output.write(
+                result_str.format(
+                    status="WARNING",
+                    file=py_file.relative_to(repo_path),
+                    line="",
```

Comment:
You might need to specify a line for the annotation to work... it might be safe to hardcode 1 here, or (ideally) figure out which line actually has the `self.id` declaration on it

Reply:
I'll push a break commit tomorrow to see if this works.

Reply:
It looks like this works correctly. I am wondering, though, if it would make sense to change the regex to try to parse the actual Python tracebacks. I've got some additional work I want to do around the infodumps now that I see how this all looks in GHA, so this isn't ready yet.